torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -320,6 +361,12 @@ def __init__(
         self.cpu_offload = cpu_offload or CPUOffload()
         self.backward_prefetch = backward_prefetch
         self.sharding_strategy = sharding_strategy or ShardingStrategy.FULL_SHARD
+        self.mixed_precision = mixed_precision
+        # Original buffer type (assumes all buffers in model are same dtype). In
+        # the case of mixed precision training, this is used to restore buffers
+        # to their original type (which may not be the same as that of the
+        # parameters in the model) when checkpointing.
+        self._orig_buffer_dtype = None
```

Comment:
as I understand, buffers don't necessarily have to be the same type as parameters (but they usually might be). A buffer is basically a tensor that's not a parameter, it could be something like an `int` counter which is a buffer so that checkpoint is easily supported. 

This diff has tests that support the case where buffer is a different type than parameters, and also supports the case where we cast buffers to a different precision than parameters as well. I feel like there are valid cases for both of these. 

Re: input/param types being float in mixed precision config, is this something we actually want to enforce? It probably won't be common but what if there are some users who want to try mixed precision in int16/int8 precision? I don't know if we should explicitly forbid this.

Reply:
that makes sense! yeah, generally the API should support all possible combinations, it is just that adding some restrictions at the beginning makes things easier to debug in the future, but if tests cases cover all combinations, it will be more confident not adding restrictions. It is up to you though